relay-server/src/actors/outcome.rs

```diff
@@ -508,61 +528,109 @@ impl HttpOutcomeProducer {
+                }
+            }
+        });
     }
 
-    fn send_http_message(&mut self, message: TrackRawOutcome, context: &mut Context<Self>) {
+    fn send_http_message(&mut self, message: TrackRawOutcome) {
         relay_log::trace!("Batching outcome");
         self.unsent_outcomes.push(message);
 
         if self.unsent_outcomes.len() >= self.config.outcome_batch_size() {
-            if let Some(pending_flush_handle) = self.pending_flush_handle {
-                context.cancel_future(pending_flush_handle);
-            }
-
-            self.send_batch(context)
-        } else if self.pending_flush_handle.is_none() {
-            self.pending_flush_handle =
-                Some(context.run_later(self.config.outcome_batch_interval(), Self::send_batch));
+            self.send_batch();
+        } else if let Either::Left(_) = &self.pending_flush_handle {
+            let timeout = self.config.outcome_batch_interval();
+            self.pending_flush_handle = Either::Right(Box::pin(tokio::time::sleep(timeout)));
         }
     }
+
+    fn handle_message(&mut self, message: HttpOutcomeProducerMessages) {
+        match message {
+            HttpOutcomeProducerMessages::TrackRawOutcome(msg) => self.send_http_message(msg),
+        };
+    }
 }
 
-impl Actor for HttpOutcomeProducer {
-    type Context = Context<Self>;
+impl Service for HttpOutcomeProducer {
+    type Messages = HttpOutcomeProducerMessages;
 }
 
-impl Handler<TrackRawOutcome> for HttpOutcomeProducer {
-    type Result = Result<(), OutcomeError>;
+impl ServiceMessage<HttpOutcomeProducer> for TrackRawOutcome {
+    type Response = ();
 
-    fn handle(&mut self, msg: TrackRawOutcome, ctx: &mut Self::Context) -> Self::Result {
-        self.send_http_message(msg, ctx);
-        Ok(())
+    fn into_messages(
+        self,
+    ) -> (
+        HttpOutcomeProducerMessages,
+        oneshot::Receiver<Self::Response>,
+    ) {
+        let (tx, rx) = oneshot::channel();
+        tx.send(()).ok();
+        (HttpOutcomeProducerMessages::TrackRawOutcome(self), rx)
     }
 }
 
+#[derive(Debug)]
+enum HttpOutcomeProducerMessages {
+    TrackRawOutcome(TrackRawOutcome),
+}
+
 struct ClientReportOutcomeProducer {
     flush_interval: Duration,
     unsent_reports: BTreeMap<Scoping, Vec<ClientReport>>,
+    // Sleep must be `Box<Pin>` in order for it to be Unpin.
+    // See: https://docs.rs/tokio/1.21.0/tokio/time/struct.Sleep.html#examples
+    flush_handle: Either<future::Pending<()>, Pin<Box<tokio::time::Sleep>>>,
+}
+
+impl Service for ClientReportOutcomeProducer {
+    type Messages = ClientReportOutcomeProducerMessages;
 }
 
 impl ClientReportOutcomeProducer {
+    pub fn start(mut self) -> ServiceAddr<Self> {
+        let (tx, mut rx) = mpsc::unbounded_channel::<ClientReportOutcomeProducerMessages>();
+        if self.flush_interval > Duration::ZERO {
+            self.flush_handle = Either::Right(Box::pin(tokio::time::sleep(self.flush_interval)));
```

Comment:
This interval is created once but never restarted. Similar to `HttpOutcomeProducer` and `OutcomeAggregator`, it would be better to apply just one strategy everywhere: Create the flush delay on demand when pushing a client report and reset it in `flush`.